cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/transforms/StreamToStructuredRecordTransform.java

```diff
@@ -0,0 +1,106 @@
+import co.cask.cdap.templates.etl.api.Transform;
+import co.cask.cdap.templates.etl.transforms.formats.RecordFormats;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import org.apache.hadoop.io.LongWritable;
+
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Transforms {@link StreamEvent} to {@link StructuredRecord}
+ */
+public class StreamToStructuredRecordTransform extends Transform<LongWritable, StreamEvent,
+                                                                 LongWritable, StructuredRecord> {
+  private static final String SCHEMA = "schema";
+  private static final String SCHEMA_TYPE = "schemaType";
+  private static SchemaWrapper schemaWrapper = null;
+
+  @Override
+  public void configure(StageConfigurer configurer) {
+    configurer.setName(StreamToStructuredRecordTransform.class.getName());
+    configurer.setDescription("Transforms a StreamEvent to StructuredRecord.");
+    configurer.addProperty(new Property(SCHEMA, "The schema of the body of stream events", true));
+    configurer.addProperty(new Property(SCHEMA_TYPE, "Type of the Schema ex: CSV, TSV etc.", true));
+  }
+
+  @Override
+  public void transform(@Nullable LongWritable inputKey, StreamEvent streamEvent,
+                        Emitter<LongWritable, StructuredRecord> emitter) throws Exception {
+
+    if (schemaWrapper == null) {
+      // read the stream event body as a StructuredRecord
+      Schema streamBodySchema = Schema.parseJson(getContext().getRuntimeArguments().get(SCHEMA));
+      schemaWrapper = getSchemaWrapper(streamBodySchema);
+    }
+
+    StructuredRecord streamEventRecord = schemaWrapper.format.read(streamEvent);
+
+    // build the structured record with timestamp, headers and copying all the fields from stream event
+    StructuredRecord.Builder builder = StructuredRecord.builder(schemaWrapper.streamSchema);
+    builder.set("ts", streamEvent.getTimestamp());
+    builder.set("headers", streamEvent.getHeaders());
+
+    for (Schema.Field field : schemaWrapper.streamBodySchema.getFields()) {
+      String fieldName = field.getName();
+      builder.set(fieldName, streamEventRecord.get(fieldName));
+    }
+
+    emitter.emit(inputKey, builder.build());
+  }
+
+  private SchemaWrapper getSchemaWrapper(Schema streamBodySchema) throws Exception {
+    // create the new schema with timestamp and headers to reatin them in StructuredRecord
+    List<Schema.Field> fields = Lists.newArrayList(
+      Schema.Field.of("ts", Schema.of(Schema.Type.LONG)),
+      Schema.Field.of("headers", Schema.mapOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.STRING))));
+    fields.addAll(streamBodySchema.getFields());
+
+    RecordFormat<StreamEvent, StructuredRecord> format = RecordFormats.createInitializedFormat(
+      new FormatSpecification(getContext().getRuntimeArguments().get(SCHEMA_TYPE), streamBodySchema,
+                              ImmutableMap.<String, String>of()));
```

Comment:
we should have some way of passing in format settings. Perhaps any setting that is prefixed by "format.setting.<key>"?